mountinfo/mounted_linux.go

```diff
@@ -0,0 +1,73 @@
+package mountinfo
+
+import (
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+)
+
+func mountedByOpenat2(path string, st *unix.Stat_t) (bool, error) {
+	// This method works for all mounts (incl. bind mounts),
+	// but requires a recent (v5.6+) linux kernel.
+	var dir, file string
+
+	switch st.Mode & unix.S_IFMT {
+	case unix.S_IFDIR:
+		dir = path
+		file = ".."
+	default:
+		dir, file = filepath.Split(path)
+	}
+
+	dirfd, err := unix.Open(dir, unix.O_PATH|unix.O_CLOEXEC, 0)
```

Comment:
This is more of a question than a requested change, but you care about symlinks in the path? If you don't want any symlinks in the path you could open `dirfd` with `openat2(RESOLVE_NO_SYMLINKS)` -- but I'm not sure what semantics you want for this function.

Reply:
I would like it to work with any paths, so let this stay as is.

And yet it makes sense to use `openat2()` here since that way we'll bail out earlier if it's not supported.